### DISCUSSION TITLE
adding ability to read /set IR and night mode to web

### DISCUSSION THREAD

**Discussion by virmaior** [gtxaspec/wz_mini_hacks#577] [Development]
Setting is easy: `cmd irled on` or off or auto can **set** the state of the IR LED.

I haven't figured out reading yet.

Are the values at GPIO49 and GPIO50 as per OPEN IPC (https://github.com/OpenIPC/firmware/blob/34ee9d06a3a6ddc1f9552faa367e55f6c6a0f490/general/package/ingenic-osdrv-t31/files/script/ircut_demo) ? 
atomcam_tools and the notes.md here indicate GPIO 49 or possibly 25.

I tried exposing GPIO 49 and reading/writing the value. I believe GPIO 49 controls whether to use the IR light when night mode is on.

(specifically, I'd like to be able to chain their behavior as I have multiple cameras pointing at the same location and they don't make consistent judgments about when to go into night mode -- they also confuse each other).

**Comment by gtxaspec** [comment]
the GPIO's for the IR leds should already be exposed via sysfs, you just have to flip them on or off via changing the value.   (actually any led's on the devices can be turned on or off via sysfs) they vary by model, I dont recall what numbers they are at the moment.  once we figure out what they are you can just read the 'value' of the gpio node 
update:
check https://github.com/gtxaspec/wz_mini_hacks/blob/master/documentation/notes.md

**Comment by virmaior** [comment]
As far as I can tell that information is wrong,

that suggests
`gpio_request lable = night_gpio gpio = 49`


night mode:
```
tail gpio38/value 1
tail gpio39/value 1
tail gpio47/value 0
tail gpio49/value 0
tail gpio50/value 0
tail gpio51/value 1
tail gpio52/value 1
tail gpio53/value 1
tail gpio62/value 0
```

different one: tail /sys/devices/virtual/gpio/*/value
```
==> /sys/devices/virtual/gpio/gpio38/value <==
1

==> /sys/devices/virtual/gpio/gpio39/value <==
1

==> /sys/devices/virtual/gpio/gpio47/value <==
1

==> /sys/devices/virtual/gpio/gpio50/value <==
0

==> /sys/devices/virtual/gpio/gpio51/value <==
1

==> /sys/devices/virtual/gpio/gpio52/value <==
1

==> /sys/devices/virtual/gpio/gpio53/value <==
1

==> /sys/devices/virtual/gpio/gpio62/value <==
0
```

vs  not night:
```
tail gpio38/value 1
tail gpio39/value 1
tail gpio47/value 0
tail gpio50/value 0
tail gpio51/value 1
tail gpio52/value 0
tail gpio53/value 0
tail gpio62/value  0


```

assuming I can read GPIO that are set to out then it looks like it's 52 & 53 for night vision mode (probably IR LED + lens cover)

(interestingly atomcam has a different GPIO for the wifi card )

they do change to the correct values when the camera switches modes
but manipulating by 
`echo "1" > gpio52/value ` and `echo "1" > gpio53/value` seems to have no effect even though both are direction out

**Comment by virmaior** [comment]
okay, some progress here.

```
Reading 47 =  see whether LED is on or off
Reading 52/53 = see whether camera is in night mode (i.e. whether it has the filter on and is displaying b/w)
```

what cannot be done is setting 52/53 directly from here. must use `cmd irled on/off/auto` regardless of the GPIO in/out state.

mostly I'm planning to use this in /cgi-bin/status.cgi and then enable coordination of night mode states for multicam.

**Comment by virmaior** [comment]
47 does seem to be whether the IR LED is on or not 
but 52/53 don't seem to match what I was hoping for

@gtxaspec What's reading the other side of `wz_mini_hacks/SD_ROOT/wz_mini/etc/rc.d/S14nightdrop` 's `/opt/wz_mini/tmp/.nd` ?
I couldn't find any reference to .nd anywhere else in the code base... but that seems to know whether we're in night mode or not. Is it a hack on iCamera?

**Reply by gtxaspec** [reply]
.nd is set in the libcallback library sources.

52/53 control the IR filter.

which device specifically are you testing on?

**Comment by virmaior** [comment]
I'm trying to figure out how to detect if the camera is IR night vision mode on a V3.

**Reply by gtxaspec** [reply]
you mean if the IR filter has been switched ON?

**Reply by gtxaspec** [reply]
you can also grep this in a script:
```
cat /proc/jz/isp/isp-m0 
****************** ISP INFO **********************
Software Version : H20211130a
SENSOR NAME : gc2053
SENSOR OUTPUT WIDTH : 1920
SENSOR OUTPUT HEIGHT : 1080
ISP OUTPUT FPS : 30 / 1
SENSOR OUTPUT RAW PATTERN : RGGB
ISP Top Value : 0xb5742249
ISP Runing Mode : Day
```
isp running mode will switch to night when its night mode

**Reply by virmaior** [reply]
that was exactly what I needed (though the `runing` typo messed with my grep for a bit)

My cameras are extremely verbose here:
```

****************** ISP INFO **********************
Software Version : H20200506a
SENSOR NAME : gc2053
SENSOR OUTPUT WIDTH : 1920
SENSOR OUTPUT HEIGHT : 1080
ISP OUTPUT FPS : 20 / 1
SENSOR OUTPUT RAW PATTERN : RGGB
ISP Top Value : 0xb5742249
ISP Runing Mode : Day
ISP WDR Mode : Disable
SENSOR Integration Time : 1406 lines
SENSOR Max Integration Time : 1683 lines
SENSOR analog gain : 2
MAX SENSOR analog gain : 205
SENSOR digital gain : 0
MAX SENSOR digital gain : 0
ISP digital gain : 1
MAX ISP digital gain : 64
ISP Tgain DB : 0
ISP EV value: 1558
ISP EV value log2: 695043
ISP EV value us: 41671
ISP EV min int: 1
ISP EV min again: 1024
ISP WB weighted rgain: 381
ISP WB weighted bgain: 601
ISP WB color temperature: 4469
ISP AWB Start rgain 230: bgain 284
Saturation : 128
Saturation : 128
Sharpness : 128
Contrast : 128
Brightness : 128
Antiflicker : 2
debug : 39262080,0,0,0,0,0,0
debug1 : 0,0,900


```

**Comment by virmaior** [comment]
okay implemented in https://github.com/gtxaspec/wz_mini_hacks/pull/627
